pytorch_lightning/overrides/distributed.py

```diff
@@ -104,6 +106,103 @@ def __iter__(self) -> Iterator[List[int]]:
         return iter(indices)
 
 
+class _DatasetFromSampler(Dataset):
+    """Dataset to create indexes from `Sampler`.
+
+    Adapted from https://github.com/catalyst-team/catalyst/blob/v22.04/catalyst/data/dataset.py#L6
+
+    Args:
+        sampler: The sampler.
+    """
+
+    def __init__(self, sampler: Union[Sampler, Iterable]) -> None:
+        if not isinstance(sampler, Sized):
+            raise MisconfigurationException(
+                "You seem to have configured a sampler in your DataLoader which"
+                " does not provide `__len__` method. The sampler was about to be"
+                " replaced by `DistributedSamplerWrapper` since `replace_sampler_ddp`"
+                " is True and you are using distributed training. Either provide `__len__`"
+                " method in your sampler, remove it from DataLoader or set `replace_sampler_ddp=False`"
+                " if you want to handle distributed sampling yourself."
+            )
+        if len(sampler) == float("inf"):
+            raise MisconfigurationException(
+                "You seem to have configured a sampler in your DataLoader which"
+                " does not provide finite `__len__` method. The sampler was about to be"
+                " replaced by `DistributedSamplerWrapper` since `replace_sampler_ddp`"
+                " is True and you are using distributed training. Either provide `__len__`"
+                " method in your sampler which returns a finite number, remove it from DataLoader"
+                " or set `replace_sampler_ddp=False` if you want to handle distributed sampling yourself."
+            )
+            raise TypeError("The given sampler must be bounded.")
+        self._sampler = sampler
+        self._sampler_list: Optional[List[Any]] = None
+
+    def __getitem__(self, index: int) -> Any:
+        """Gets element of the dataset.
+
+        Args:
+            index: index of the element in the dataset
+        Returns:
+            Single element by index
+        """
+        if self._sampler_list is None:
+            self._sampler_list = list(self._sampler)
```

Comment:
why not initialize it inside init?

Reply:
Because it might be an expensive operation. And, if you only want `__len__`, it is not necessary to have this.